debian/manpages/ruby-prof.1
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "RUBY\-PROF" "1" "December 2013" "" ""
.
.SH "NAME"
\fBruby\-prof\fR \- A fast ruby profiler
.
.SH "SYNOPSIS"
\fBruby\-prof\fR \fIoptions\fR <script\.rb> [\-\-] [profiled\-script\-command\-line\-options]
.
.SH "DESCRIPTION"
\fBruby\-prof\fR is a fast profiler for ruby\. It can be used as a stand\-alone program (that this man page documents) or as a ruby library\. Simply run the ruby program you want to profile using ruby\-prof : \fBruby\-prof script\.rb\fR At the end of execution the time spent in each method is displayed\.
.
.SH "OPTIONS"
.
.TP
\fB\-p\fR, \fB\-\-printer=printer\fR
Select a printer:
.
.br
flat \- Prints a flat profile as text (default)\.
.
.br
flat_with_line_numbers \- same as flat, with line numbers\.
.
.br
graph \- Prints a graph profile as text\.
.
.br
graph_html \- Prints a graph profile as html\.
.
.br
call_tree \- format for KCacheGrind
.
.br
call_stack \- prints a HTML visualization of the call tree
.
.br
dot \- Prints a graph profile as a dot file
.
.br

.
.TP
\fB\-m\fR, \fB\-\-min_percent=min_percent\fR
The minimum percent a method must take before being included in output reports\. this option is not supported for call tree\.
.
.TP
\fB\-f\fR, \fB\-\-file=path\fR
Output results to a file instead of standard out\.
.
.TP
\fB\-\-mode=measure_mode\fR
Select what ruby\-prof should measure:
.
.br
process \- Process time (default)\.
.
.br
wall \- Wall time\.
.
.br
cpu \- CPU time (Pentium and PowerPCs only)\.
.
.br
allocations \- Object allocations (requires patched Ruby interpreter)\.
.
.br
memory \- Allocated memory in KB (requires patched Ruby interpreter)\.
.
.br
gc_runs \- Number of garbage collections (requires patched Ruby interpreter)\.
.
.br
gc_time \- Time spent in garbage collection (requires patched Ruby interpreter)\.
.
.br

.
.TP
\fB\-s\fR, \fB\-\-sort=sort_mode\fR
Select how ruby\-prof results should be sorted:
.
.br
total \- Total time
.
.br
self \- Self time
.
.br
wait \- Wait time
.
.br
child \- Child time
.
.br

.
.TP
\fB\-\-replace\-progname\fR
Replace $0 when loading the \.rb files\.
.
.TP
\fB\-v\fR
Show version, set $VERBOSE to true, profile script if option given
.
.TP
\fB\-d\fR
Set $DEBUG to true
.
.TP
\fB\-R\fR, \fB\-\-require\-noprof lib\fR
require a specific library (not profiled)
.
.TP
\fB\-E\fR, \fB\-\-eval\-noprof code\fR
execute the ruby statements (not profiled)
.
.TP
\fB\-x\fR, \fB\-\-exclude regexp\fR
exclude methods by regexp (see method elimination)
.
.TP
\fB\-X\fR, \fB\-\-exclude\-file file\fR
exclude methods by regexp listed in file (see method elimination)
.
.TP
\fB\-\-exclude\-common\-cycles\fR
make common iterators like Integer#times appear inlined
.
.TP
\fB\-\-exclude\-common\-callbacks\fR
make common callbacks invocations like Integer#times appear inlined so you can see call origins in graph
.
.TP
\fB\-h\fR, \fB\-\-help\fR
Show help message
.
.TP
\fB\-\-version\fR
Show version 0\.13\.1
.
.SH "COPYRIGHT"
ruby\-prof was written by Shugo Maeda \fIshugo@ruby\-lang\.org\fR\.
.
.P
This manual page was written by Arnaud Cornet \fIarnaud\.cornet@gmail\.com\fR and updated by Jonas Genannt \fIjonas\.genannt@capi2name\.de\fR for the Debian project (but may be used by others)\.
.
.SH "SEE ALSO"
The ruby\-prof website, https://github\.com/ruby\-prof/ruby\-prof


CHANGES
1.3.1 (2020-03-11)
=====================
* Add max_percent and filter_by options to flat printer (Sean McGivern)
* Include binary in mingw64 build (Charlie Savage)

1.3.0 (2020-02-22)
=====================
* Update C code to use newer RTypedData API versus older RData API.
* Update C code to use rb_cObject versus the older, deprecated, rb_cData.
* Update memory management - CallInfo's now keep alive their owning Profile instances. Fixes crashes that
  could happen in rare instances.

1.2.0 (2020-01-23)
=====================
* Fix call stack printer broken in version 1.0.0 (Charlie Savage)
* Ruby 2.7.0 support (Charlie Savage)
* Switch to Windows high precision timer for measuring wall time (Charlie Savage)
* Much better support for reverse call trees - if you start RubyProf at the bottom
  of a call tree it will correctly calculate times as it ascends the tree (Charlie Savage)
* Port AggregateCallInfo from Ruby to C to increase performance
* API change - CallInfo has been renamed CallTree
* Fix crashes on Ruby 2.4.*

1.1.0 (2019-12-14)
=====================
Changes

* Reimplement merge_fibers (Charlie Savage)
* Fix crash caused by threads being freed when profiles are freed (Charlie Savage)

1.0.0 (2019-07-29)
=====================
ruby-prof's development stretches all the way back to 2005. Fourteen years later, it seems time for version 1.0!
Version 1.0 is a significant update that greatly improves the internals of ruby-prof and adds a number of improvements.

Changes (Charlie Savage):

* Profiling is significantly faster - 5x in some cases
* Recursive profiles are finally handled correctly.  Yeah!!!
* Redesigned reports (Chirs Whitefield)
* New documentation website (https://ruby-prof.github.io)
* The ability to measure allocations and memory usage using a standard (unpatched) version of ruby
* The ability to save and reload profiling results for later analysis
* The ability track object allocations

At the same time, it was also time to remove old code and deprecated methods. These changes include (Charlie Savage):

* Drop support for old versions of ruby. Currently 2.4 and up are required.
* Remove support for method elimination (instead use method exclusion).
* Remove the CPU_TIME measurement mode since it duplicates the PROCESS_TIME mode and required inline assembly code
* Remove the GC_TIME and GC_RUNS measurement modes since they required patched ruby versions. In modern ruby it is possible to reimplement this functionality, but it does not seem of that much value (shout out if you want it back).
* Merge the RubyProf::FlatPrinterWithLineNumbers printer into RubyProf::FlatPrinter.
* Removed support for RUBY_PROF_EDITOR_URI environment variable that specified the link scheme. This features was more confusing then helpful.


0.18.0 (2019-05-27)
=====================
Fixes
------
* Fix grouping by thread/fiber (Charlie Savage)
* Fix include/exclude thread checking (Charlie Savage)
* Fix failing tests (Charlie Savage)
* Fix grouping by thread_id and fiber_id (Charlie Savage)
* Update Visual Studio solution to 2019 (Charlie Savage)
* Adjust AbstractPrinter#editor_uri so it's possible to disable uri (Maciek Dubiński)-
* Fix -X/--exclude-file option not working (John Lindgren)
* Different directory for middleware output in readme (Aldric Giacomoni)
* Fixing spelling mistakes and typos (David Kennedy and Roger Pack)
* Add "allow_exceptions" option to Profile (Ryan Duryea)
* Update readme (Jan Piotrowski, Nathan Seither, victor-am)
* Travis updates (Niocals Leger and Charlie Savage)
* Fix NewRelic module detection in exclude_common_methods! (Ben Hughes)

0.17.0 (2017-12-17)
=====================
* Changed method/class names produced by the callgrind printer to be more kcachegrind
  friendly (thx to Nelson Gauthier and Ben Hughes, see https://github.com/ruby-prof/ruby-prof/pull/200).
* Recursive instances of a method call are now dtected during profiling (thx to Nelson Gauthier and
  Ben Hughes, see https://github.com/ruby-prof/ruby-prof/pull/201).
* Methods can now be ignored during a profiling run instead of eliminating them
  after a profile run, which has been deprecated (thx to Nelson Gauthier and Ben Hughes,
  see https://github.com/ruby-prof/ruby-prof/pull/202).
* Use RUBY_PROF_EDITOR_URI=emacs to open source code links in the one true editor.
  (thx Alexander Egorov, see https://github.com/ruby-prof/ruby-prof/pull/183 and
  https://github.com/typester/emacs-handler).
* Fixed that calling profile.stop would stop profiling all active profiling runs.
* Fixed that remembering parent call frame pointer before stack reallocation could cause a segfault.

0.16.2 (2016-08-25)
=====================
* fixed incorrect setting of RUBY_VERSION macro

0.16.1 (2016-08-24)
=====================
* fixed compile problem with ruby 2.1.10

0.16.0 (2016-08-24)
=====================
* ruby-prof can now be installed if ruby has been compiled with --disable-gems or
  RUBYOPT includes --disable-gems (thx to Arthur Nogueira Neves)
* Profile.new now accepts and prefers an options hash:
  - measure_mode: one of the RubyProf measure mode constants
  - exclude_threads: an array of threads to ignore when collecting profile data
  - include_threads: an array of threads to include when collecting profile data
  - merge_fibers: whether or not to merge the profile data for all fibers under their creating thread
* overhauled printer for cachegrind profiles:
  - profile data for each thread dumped to a separate file, as kcachegrind does not
    fully implement the cachegrind specification
  - this required a change in interface: it's now necessary to specify a directory
    where to dump the cachegrind files
  - naming convention for cachegrind files changed so that kcachegrind shows them
    all in the open dialog
* improved Rack middleware:
  - it's now possible to use the MultiPrinter

lib/ruby-prof.rb
# encoding: utf-8

# Load the C-based binding.
begin
  RUBY_VERSION =~ /(\d+\.\d+\.\d+)/
  require "#{$1}/ruby_prof.so"
rescue LoadError
  require "ruby_prof.so"
end

require 'ruby-prof/version'
require 'ruby-prof/call_tree'
require 'ruby-prof/compatibility'
require 'ruby-prof/measurement'
require 'ruby-prof/method_info'
require 'ruby-prof/profile'
require 'ruby-prof/rack'
require 'ruby-prof/thread'

module RubyProf
  autoload :CallTreeVisitor, 'ruby-prof/call_tree_visitor'
  autoload :AbstractPrinter, 'ruby-prof/printers/abstract_printer'
  autoload :CallInfoPrinter, 'ruby-prof/printers/call_info_printer'
  autoload :CallStackPrinter, 'ruby-prof/printers/call_stack_printer'
  autoload :CallTreePrinter, 'ruby-prof/printers/call_tree_printer'
  autoload :DotPrinter, 'ruby-prof/printers/dot_printer'
  autoload :FlatPrinter, 'ruby-prof/printers/flat_printer'
  autoload :GraphHtmlPrinter, 'ruby-prof/printers/graph_html_printer'
  autoload :GraphPrinter, 'ruby-prof/printers/graph_printer'
  autoload :MultiPrinter, 'ruby-prof/printers/multi_printer'

  # :nodoc:
  # Checks if the user specified the clock mode via
  # the RUBY_PROF_MEASURE_MODE environment variable
  def self.figure_measure_mode
    case ENV["RUBY_PROF_MEASURE_MODE"]
    when "wall", "wall_time"
      RubyProf.measure_mode = RubyProf::WALL_TIME
    when "allocations"
      RubyProf.measure_mode = RubyProf::ALLOCATIONS
    when "memory"
      RubyProf.measure_mode = RubyProf::MEMORY
    when "process", "process_time"
      RubyProf.measure_mode = RubyProf::PROCESS_TIME
    else
      # the default is defined in the measure_mode reader
    end
  end
end

RubyProf::figure_measure_mode
